iib/workers/tasks/build_regenerate_bundle.py

```diff
@@ -363,14 +376,65 @@ def _resolve_image_pull_specs(bundle_metadata, labels, pinned_by_iib, registry_r
             log.debug('%s will be replaced with %s', pullspec, new_pullspec.to_str())
             replacement_pullspecs[pullspec] = new_pullspec
 
+    _replace_csv_pullspecs(bundle_metadata, replacement_pullspecs, set_related_images=True)
+
+
+def _replace_image_name_from_labels(bundle_metadata, replacement_template):
+    """
+    Replace repo/image-name in the CSV pull specs with values from their labels.
+
+    :param dict bundle_metadata: the dictionary of CSV's and relatedImages pull specifications
+    :param str replacement_template: the template specifying which label values to use for
+        replacement
+    """
+    replacement_pullspecs = {}
+    label_regex = r'\{(.*?)\}'
+    replacement_labels = re.findall(label_regex, replacement_template)
+
+    for pullspec in bundle_metadata['found_pullspecs']:
+        new_pullspec = ImageName.parse(pullspec.to_str())
+        pullspec_labels = get_image_labels(pullspec.to_str())
+        modified_namespace_repo = replacement_template
+
+        for label in replacement_labels:
+            label_value = pullspec_labels.get(label)
+            if label_value:
+                modified_namespace_repo = re.sub(
+                    r'\{%s\}' % label, label_value, modified_namespace_repo
+                )
+
+        invalid_labels = re.findall(label_regex, modified_namespace_repo)
+        if invalid_labels:
+            raise IIBError(
+                f'Pull spec {pullspec.to_str()} is missing the following label(s)'
+                f' required in the image_name_from_labels {replacement_template}:'
+                f' {", ".join(invalid_labels)}'
+            )
+        new_pullspec.namespace, new_pullspec.repo = modified_namespace_repo.split('/', 1)
+        replacement_pullspecs[pullspec] = new_pullspec
+
+    _replace_csv_pullspecs(bundle_metadata, replacement_pullspecs)
```

Comment:
Why not `set_related_images` ?

Reply:
Because we already `set_related_images` when resolving the pull_specs. If we do it again here, operator-manifest adds an another entry instead of replacing the existing one. Do you think this should be changed in the operator-manifest library itself?

Reply:
Ok that makes sense. Maybe add a comment?